nifi-registry-provider-api/src/main/java/org/apache/nifi/registry/flow/FlowProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.registry.flow;
+
+import org.apache.nifi.registry.provider.Provider;
+
+/**
+ * A service that can store and retrieve versioned flow snapshots.
+ *
+ * NOTE: Although this interface is intended to be an extension point, it is not yet considered stable and thus may
+ * change across releases until the registry matures.
+ */
+public interface FlowProvider extends Provider {
```

Comment:
Kevin, thanks for taking a look... 

I originally had shortened the name from FlowPersistenceProvider to FlowProvider to make some of the supporting classes names less verbose, but with the way things turned out I think it is a non-issue now and we can go back to calling it FlowPersistenceProvider. I'll make that change and update the PR.

I envision there being a service above the providers, for now I'm calling it RegistryService, which would coordinate the operations... So for example, there would be a "snapshot method" on the service which would do something like the following:
-  Create any necessary versions or ids
-  Call the metadata provider to persist the snapshot metadata
-  Call a serializer to serialize the snapshot
-  Call the flow persistence provider to save the serialized bytes

For the search example, we probably have to think through what we want to support searching on. Searching by name is more like searching on metadata, where as if someone wants to find a snapshot that contains processor XYZ, then that would have to be searching into the content of each snapshot. I could potentially see having a "search provider" that separated search into its own thing, but need to think through it more.